### ISSUE TITLE
How do master and develop branches look in Gitflow?

### ISSUE THREAD

**Issue by martyychang** [martyychang/git-docs#15]
Given a hypothetical situation with two teams working in parallel, both on two-week sprints with joint deployments to production at the end of each sprint, as a release manager how can I support three environments and a branching strategy where shared branches correctly mirror each environment?

Environment | Branch | Description
----- | ----- | -----
Production | `master` | Self-explanatory
Preproduction | `preprod` | The environment where final acceptance testing and end-to-end testing is performed, in order to decide whether stories are "done" to the point of being production-ready
Staging | `staging` | This is analogous to `develop` in [the Gitflow setup][1], where developers merge and deployed approved product increments via pull requests

[1]: https://www.atlassian.com/git/tutorials/comparing-workflows/gitflow-workflow

A few wrinkles in the current setup. Let's say we have two teams, Alpha Team and Bravo Team.

* Alpha Team allows team members to test and validate each others work. The product owner does not have to be involved in final acceptance of a story.
* Bravo Team requires product owner acceptance or by trusted proxies for the product owner. These testers are _NOT_ regular team members.

Historically, in order to facilitate testing by the product owner in a stable environment while allowing for fixes within the sprint, code freeze was instituted on the shared `staging` branch at the end of the last workday before testing was to commence. Acceptance testing for Bravo Team's work occurred on a Monday, with Tuesday as a buffer day in case additional testing was needed. Tuesday was the official end of both sprints for Alpha Team and for Bravo Team, and accepted work across both teams for the sprint ending on the Tuesday was then deployed on Wednesday.

One reason for the code freeze was the desire to maintain official history of development in the Git repo. The way development currently works maintains the following truths.

* `master` is never ahead of `preprod` or `staging`
* At the start of a new sprint, `master` is made flush with both `preprod` and `staging`. This is done either by offline `git checkout preprod && git merge master` or deleting then recreating the branch on GitHub.

This mode of development looks "clean" as there are never growing divergences reported by GitHub as more and more disparate merge commits are accumulated on each of the shared branches.

However, this rigid structure meant that to avoid divergence of branch histories `staging` could not be touched during the 2-day testing period and the lag between end of sprint and the associated deployment. This created many frustrations among team members and also with Alpha Team's product owner.

Wanting to eliminate the code freeze and support a more continuous development workflow, the following questions should be answered.

* What do `master` and `staging` look like after two deployments?
** For testing purposes, create a new commit on `staging` followed by a new commit on `preprod` followed by another commit on `staging` before merging into `master`
** Try **Squash and merge** from `preprod` into `master`
** Try **Merge pull request** into `preprod` into `master`
* Is there an application for "[Rebase and merge][2]" in avoiding code freeze?

[2]: https://blog.github.com/2016-09-26-rebase-and-merge-pull-requests/

**Comment by martyychang** [comment]
Interesting... so it seems after merging my `release` branch into `master`, I cannot merge `master` easily into `develop`. Screenshot below showing the warning from GitHub.

![image](https://user-images.githubusercontent.com/6706834/50066544-7b76ab00-018a-11e9-9645-c4567f7cfe8d.png)

**Comment by martyychang** [comment]
Looks like the conflict may be specific to the way I edited the file, modifying the last line in the README. Needless to say merge conflicts may arise in this mode of development, although let's see if the merge proceeds more smoothly if I do not have any conflicts.

![image](https://user-images.githubusercontent.com/6706834/50066656-11aad100-018b-11e9-9fa4-8653d7712a9d.png)

**Comment by martyychang** [comment]
Well, the good news is that the convergent history does still appear to be preserved along with _all_ verified commits, based on the way the merge conflict was resolved using GitHub.

![image](https://user-images.githubusercontent.com/6706834/50066726-7b2adf80-018b-11e9-96b7-36ffdd53a583.png)

The history looks a bit strange but I suppose the team can get used to it.

**Comment by martyychang** [comment]
I think I have the answer I need to confidently move forward with this new workflow, without exhausting the other options. To summarize what I've observed here.

* History of merges along with pull requests and approvals is preserved
* Verified commits preserved and enforced
* History between `master` and `staging` can be made convergent after each deployment (i.e., merge into `master`). This will require a bit more thoughtful use of `preprod` perhaps or some other transient branch, to avoid having `master` deviate from `staging`.

So, conventionally if the plan is to follow source-driven development and have branches mirror environments, after a deployment to **production** the following should be true.

* `master` matches **Production**
* `preprod` matches **Preproduction**
* `staging` is convergent with both `master` and `preprod` and will have at least one additional merge commit from the reconciliation

**Comment by martyychang** [comment]
For historical reference, I've kept the branches I used for this spike.

* `master-15`
* `develop-15`
* `deploy/2018.12.17.421`
